### DISCUSSION TITLE
Trouble with the query API - no permission

### DISCUSSION THREAD

**Discussion by robert-lill-tie** [alapierre/ksef-java-rest-client#98] [Q&A] [answered]
Hi,
I'm trying to get the examples to work in the TEST environment.
(/ksef-sample/src/main/java/io/alapierre/ksef/sample/Main.java)
I can upload inbound and outbound invoices, I can see them in the portal.
I can even download that same invoice, but I fail to list the invoices to be downloaded.

Whatever I change in the example code leads to a JSON-format error, so I leave everything as it is.
But still the query fails and the response has the exceptionDescription: Brak uwierzytelnienia

Authentication worked for uploading, so I think it's fine. And in the portal I have provided grants
(Wystawianie faktur, Dostęp do faktur) and generated my accessToken.
What am I missing here?

**Comment by robert-lill-tie** [comment]
Additional finding:
If I send the request with the same sessionToken using Postman
and a manually crafted query JSON, I get a nice result.
`{
    "queryCriteria": {
        "acquisitionTimestampThresholdFrom": "2023-11-17T07:09:27Z",
        "acquisitionTimestampThresholdTo": "2023-11-30T16:09:27Z",
        "subjectType": "subject2",
        "type": "incremental"
    }
}`
So, it seems to be something strange in the Java API code.

        InvoiceQueryRequest request = InvoiceQueryRequest.builder()
                .queryCriteria(InvoiceQueryRequest.QueryCriteria.builder()
                        .subjectType("subject2")
                        .type("incremental")
                        .acquisitionTimestampThresholdFrom(zapytaniaApi.convertDate(DateUtils.firstDayOfMonth(LocalDate.now())))
                        .acquisitionTimestampThresholdTo(zapytaniaApi.convertDate(LocalDateTime.now()))
                        .build())
                .build();

        KsefResultStream.builder(
                page -> new InvoiceQueryResponseAdapter(zapytaniaApi.invoiceQuery(sessionToken, request, 100, page)))
                .forEach(System.out::println);
    }

The response is:
`Error in API call 400 (Bad Request) exception body {"exception":{"serviceCtx":"srvTEMFC","serviceCode":"20231130-EX-870C5F98A2-52920B2A1C-B5","serviceName":"online.query.query.invoice","timestamp":"2023-11-30T08:49:44.684Z","referenceNumber":"20231130-SE-B946AE856C-1D00A25BE8-BF","exceptionDetailList":[{"exceptionCode":21304,"exceptionDescription":"Brak uwierzytelnienia."}]}}`

**Comment by robert-lill-tie** [comment, marked as answer]
I think I found the answer:
The program needs to wait between the creation of the session token and sending the query request.
If I send the query immediately or even 500ms later, I get an error. If I wait one second, I get the expected result.
Not very satisfying, but so now I have a workaround.

**Comment by alapierre** [comment]
yest, this is correct - you need wait about 5 sec. The same problem is on Demo environment. I'm using retry mechanizm and try to execute query 3 times with 5 sec delay. This issue was reported by me to Ministry of Finance couple months ago - by wit no any answer. There is a special exception type for this situation: `ksef-client-spec/src/main/java/io/alapierre/ksef/client/exception/NoAuthenticationException.java` you can check `mapHttpResponseStatusToException` ntehod in `AbstractApiClient`
